### PR TITLE
AM-1091: License notification in UI

### PR DESCRIPTION
**Pull request by mstasgravitee**
## :id: Reference related issue. 
https://gravitee.atlassian.net/browse/AM-1091

## :pencil2: A description of the changes proposed in the pull request
The information is present when the license expires in 30 days and the user has "license_notification_read". Frontend mocks are not ready at the moment. 

## :computer: Add screenshots for UI
29 days
Screenshot from 2024-01-05 09-21-53

12 days
Screenshot from 2024-01-05 09-21-46

expired
Screenshot from 2024-01-05 09-21-34

when the menu is collapsed
Screenshot from 2024-01-02 09-41-40

permission to see the license 
image

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/platform/LicenseResource.java [hunk 1]
```diff
@@ -38,21 +39,22 @@ public class LicenseResource extends AbstractResource {
...
     @Consumes(MediaType.APPLICATION_JSON)
     public void get(@Suspended final AsyncResponse response) {
         final License platformLicense = licenseManager.getPlatformLicense();
-        GraviteeLicense license;
-        license = GraviteeLicense.builder()
+        GraviteeLicense license = GraviteeLicense.builder()
                 .tier(platformLicense.getTier())
                 .packs(platformLicense.getPacks())
                 .features(platformLicense.getFeatures())
+                .expirationDate(isLicenseExpirationNotifierEnabled ? platformLicense.getExpirationDate() : null)
```
Comment (leleueri) [inline gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/platform/LicenseResource.java:55]: **suggestion**: in order to manage the notification based on the role, maybe we should provide a new method into the "permissionService" to check if a user has at least a role in a given list of role. If the user has this role, we should set the expirationDate attribute.

File: gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml [hunk 2]
```diff
@@ -223,6 +223,8 @@ notifiers:
     sslTrustAll: false
     #sslKeyStore: /path/to/keystore
     #sslKeyStorePassword: changeme
+  licenseExpiration:
+    enabled: true
```
Comment (leleueri) [inline gravitee-am-management-api/gravitee-am-management-api-standalone/gravitee-am-management-api-standalone-distribution/src/main/resources/config/gravitee.yml:227]: **thought**: I'm not convince that this setting should be here. The `notifiers` section is here to declare how notifications can be sent to a user (email, UI, ...) but this settings is more about enabling or not the license notification... :thinking: Maybe we should introduce a "license" section (or "enterprise" section) :thinking: In any case, we have to be align with APIM as much as possible as this license notification will be present in both products.

**Review by leleueri — CHANGES_REQUESTED** [review]

**Comment by leleueri** [comment]
I click on "request change" but both comments are more a discussion.

File: gravitee-am-ui/src/app/components/sidenav/sidenav.component.html [hunk 3]
```diff
@@ -39,6 +39,29 @@
         <gio-menu-item [icon]="item.icon" [active]="rla.isActive" [iconRight]="item?.iconRight$ | async">{{ item.label }}</gio-menu-item>
       </a>
     </gio-menu-list>
+    <gio-menu-footer *ngIf="showExpirationMessage() && canDisplayOrganizationSettings()">
+      <gio-menu-header>
+        <div
+          style="background-color: white; border-radius: 10px; align-self: flex-end"
```
Comment (ashraf706) [inline gravitee-am-ui/src/app/components/sidenav/sidenav.component.html:45]: wouldn't it be cleaner to add the all the styles in the css files? Also `align-self: flex-end` may not have any effect due to parent component display is set to block.

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/platform/LicenseResource.java (hunk 1 above)
Comment (mstasgravitee) [reply gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/platform/LicenseResource.java:55]: notification will be shown based a new permission.

File: gravitee-am-model/src/main/java/io/gravitee/am/model/permissions/Permission.java [hunk 4]
```diff
@@ -90,8 +90,9 @@ public enum Permission {
     APPLICATION_RESOURCE(ReferenceType.ORGANIZATION, ReferenceType.ENVIRONMENT, ReferenceType.DOMAIN, ReferenceType.APPLICATION),
     APPLICATION_ANALYTICS(ReferenceType.ORGANIZATION, ReferenceType.ENVIRONMENT, ReferenceType.DOMAIN, ReferenceType.APPLICATION),
     APPLICATION_FLOW(ReferenceType.ORGANIZATION, ReferenceType.ENVIRONMENT, ReferenceType.DOMAIN, ReferenceType.APPLICATION),
+    LICENSE_NOTIFICATION(ReferenceType.ORGANIZATION, ReferenceType.ENVIRONMENT, ReferenceType.DOMAIN, ReferenceType.APPLICATION),
+    INSTALLATION(ReferenceType.ORGANIZATION);
 
-    INSTALLATION(ReferenceType.PLATFORM);
```
Comment (leleueri) [inline gravitee-am-model/src/main/java/io/gravitee/am/model/permissions/Permission.java:94]: **issue**: please do not change this one. PLATFORM is higher than ORGANIZATION. If we have to chance something one day, it should be adding the ORGANIZATION Reference to the INSTALLATION but PLATFORM must be preserved

File: gravitee-am-ui/src/app/components/sidenav/sidenav.component.ts [hunk 5]
```diff
@@ -98,6 +128,10 @@ export class SidenavComponent implements OnInit, OnDestroy {
     return this.environments && this.environments.length > 0 && !this.router.url.startsWith('/settings');
   }
 
+  canDisplayLicence(): boolean {
+    return !this.router.url.startsWith('/settings') && this.authService.hasPermissions(['license_notification_read']);
```
Comment (leleueri) [inline gravitee-am-ui/src/app/components/sidenav/sidenav.component.ts:132]: **question**: Why are you limiting the notification to the `/settings` path ? I can't see anything into the JIRA description about this. IMO notification should alaways be displayed. wdyt ?

**Review by leleueri — CHANGES_REQUESTED** [review]

**Review by leleueri — APPROVED** [review]

**Pull request merged**